### PR TITLE
Add decomposition for permute_copy

### PR DESCRIPTION
**Pull request by rec**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130944

* Extracted from #129476

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @XilunWu

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -16953,6 +16953,22 @@ def sample_inputs_alias_copy(op_info, device, dtype, requires_grad, **kwargs):
            supports_varargs=True,
            sample_inputs_func=sample_inputs_permute,
            reference_inputs_func=reference_inputs_permute),
+    OpInfo('permute_copy',
+           ref=np.transpose,
+           dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16, torch.chalf),
+           supports_out=True,
+           assert_autodiffed=True,
+           autodiff_fusible_nodes=[],  # aliases inputs, shouldn't be fused
+           autodiff_nonfusible_nodes=[],  # aliases inputs, shouldn't be fused
```
Comment (peterbell10) [inline torch/testing/_internal/common_methods_invocations.py:16962]: This copy op shouldn't alias anything so I don't think this applies

**Comment by rec** [comment]
topic: not user facing

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (rec) [reply torch/testing/_internal/common_methods_invocations.py:16962]: Gone!

**Review by amjames — APPROVED** [review]

**Review by eellison — APPROVED** [review]

**Comment by rec** [comment]
Hey, @eellison!

This code was originally part of a [bigger decomposition pull request](https://github.com/pytorch/pytorch/pull/128416), and it caused executorch breakage at the time and was reverted.

Unfortunately, at the time we didn't figure out how to run executorch tests using the code in that commit, so I split that commit into smaller, independent commits to find which part caused the breakage. This is one of the last of them.

If I could just run a debugger on the failing executorch test using that commit ID and reproduce the problem, I'm confident I could fix it pretty fast.

Any suggestions or information would be appreciated!

**Comment by eellison** [comment]
cc @GregoryComer, blame shows you associated with the failing test, would you please advise here ?

**Comment by GregoryComer** [comment]
If I'm following correctly, the question is how to run the executorch tests? I'd recommend pulling the executorch repo and you can then run the et python tests individually. It should use the installed pytorch version, so if you build and install locally, you can then run the tests via pytest. Here's the relevant unit test workflow script in CI: https://github.com/pytorch/executorch/blob/main/.github/workflows/_unittest.yml. Feel free to correct me if I'm misunderstanding the question. Thanks!

For the XNNPACK op tests, I'll CC @mcr229, as well.

**Comment by rec** [comment]
I know how to run the executorch tests. The problem (the last time I tried this) was running them against a specific Pytorch commit ID so I can reproduce the problem.

Unfortunately, the executorch installation script installs a release version of Pytorch. Both I and @amjames tried to finagle it to use a local version of Pytorch with the commit ID in question but failed.

**Comment by eellison** [comment]
I can update the failing test to skip the check for explicit permute op when we land again. @GregoryComer, If executorch doesn't want to decompose permute copy would you prepare the corresponding changes that would opt out of the decomp?

**Comment by rec** [comment]
@eellison: All right, this seems essentially green - just one Mac test fails, apparently due to infrastructure oopsies.

So I'm going to merge it, and let's see what happens!

**Comment by clee2000** [comment]
@pytorchbot revert -m "breaking internal builds D64418214 cc @digantdesai @GregoryComer to help get this fixed and remerged" -c ghfirst

```
RuntimeError: Expected to not find "torch.ops.higher_order.executorch_call_delegate" but found it
    getitem_1 = executorch_call_delegate_1[0];  executorch_call_delegate_1 = None
    aten_clone_default = executorch_exir_dialects_edge__ops_aten_clone_default(getitem_1, memory_format = torch.contiguous_format);  getitem_1 = None
    executorch_call_delegate = torch.ops.higher_order.executorch_call_delegate(lowered_module_0, aten_clone_default);  lowered_module_0 = aten_clone_default = None
                               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    getitem = executorch_call_delegate[0];  executorch_call_delegate = None
    return (getitem,)
From CHECK-NOT: torch.ops.higher_order.executorch_call_delegate
```

**Comment by rec** [comment]
@digantdesai @GregoryComer: arg, sorry about repushing this, I would have known this was going to happen, as we never tracked down the issue last time.

There was an unrelated issue blocking a bunch of decomposition pull requests, and when I cleared that, I tried to remerge all of them, and included this one by mistake.

I feel this might be easy to debug if I could run executorch's tests against a specific commit ID for `pytorch`, instead of a production build - we tried to do that last time this came up, but failed.  I don't suppose you have instructions on how to do that?

Otherwise, I know some more tricks since last time, I could give it a go again.

**Comment by larryliu0820** [comment]
Hi @rec can you provide some context on this PR? Do you particularly care about core ATen opset (mostly used by ExecuTorch) or inductor?

**Comment by GregoryComer** [comment]
@rec I'm taking a look at the ExecuTorch breakages. To clarify, were you expecting this change to use this decomposition by default for all export?

File: torch/_decomp/__init__.py [hunk 2]
```diff
@@ -578,6 +578,7 @@ def _core_aten_decompositions_post_autograd() -> (
             aten.norm,
             aten.ones,
             aten.ones_like,
+            aten.permute_copy,
```
Comment (larryliu0820) [inline torch/_decomp/__init__.py:581]: If you only care about inductor, can you remove this line and add it https://github.com/pytorch/pytorch/blob/main/torch/_inductor/decomposition.py#L48 ?

Comment (kimishpatel) [reply torch/_decomp/__init__.py:581]: yes please

**Comment by GregoryComer** [comment]
@rec Taking a look at the test breakage, it looks like what's happening is that permute_copy gets decomposed into permute (view) + clone, then gets converted back into permute_copy + clone (no-op). This is because views are stripped from the IR at this point (at least for edge), so what ends up happening is it just adds an extra clone op. Do you care about this decomp being applied to edge?

**Comment by rec** [comment]
@GregoryComer @larryliu0820'

All right, thanks, this was extremely useful!

I now understand what's going on much better, and I took @larryliu0820's suggestion, as we really only need this decomposition for inductor right now.

I'll be sure to remember that changing `_decomp._core_aten_decompositions_post_autograd()` is something that might cause breakage downstream from `pytorch`.

Once the tests are all green, I'll push this again, believing that it won't bother you this time.

**Comment by larryliu0820** [comment]
@rec thanks for taking the time fixing it! Yeah we should definitely do a better job documenting these things, and probably capture the breakage earlier, for the sake of devX.

**Pull request closed without merging**